kaldb/src/main/java/com/slack/kaldb/server/ManagerApiGrpc.java

```diff
@@ -171,6 +179,36 @@ public void updatePartitionAssignment(
     }
   }
 
+  @Override
+  public void restoreSnapshot(
+      ManagerApi.RestoreSnapshotRequest request,
+      StreamObserver<ManagerApi.RestoreSnapshotResponse> responseObserver) {
+    try {
+      Preconditions.checkArgument(
+          request.getStartTimeEpochMs() < request.getEndTimeEpochMs(),
+          "Start time must not be after end time");
+      Preconditions.checkArgument(
+          !request.getServiceName().isEmpty(), "Service name must not be empty");
+
+      List<SnapshotMetadata> snapshotsToRestore =
+          fetchSnapshots(
+              snapshotMetadataStore.getCached(),
+              datasetMetadataStore,
+              request.getStartTimeEpochMs(),
+              request.getEndTimeEpochMs(),
+              request.getServiceName());
+
+      replicaRestoreService.queueSnapshotsForRestoration(snapshotsToRestore);
+
+      responseObserver.onNext(
+          ManagerApi.RestoreSnapshotResponse.newBuilder().setStatus("success").build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Error restoring snapshot", e);
+      responseObserver.onError(Status.UNKNOWN.withDescription(e.getMessage()).asException());
```

Comment:
If you catch the `SizeLimitExceededException` you can log that out as only an `info` level, and provide a more targeted status code in the GRPC response (perhaps `RESOURCE_EXHAUSTED` or `INVALID_ARGUMENT`? either seems like it could apply).